### PR TITLE
feat(speculative): add Qwen3-VL support for DFlash training

### PR DESCRIPTION
**Pull request by skierat**
### What does this PR do?         

  Type of change: new feature

  Adds online DFlash training support for Qwen3-VL–style vision-language models.
                                                             
  Changes include:

  - Load VLMs through the Transformers 5 `AutoModelForImageTextToText` API, while retaining compatibility with the legacy VLM auto-model API.
  - Run the base model through its top-level multimodal forward when image/video inputs are present, ensuring vision embeddings are injected before collecting DFlash target hidden states.
  - Extend `VisionLanguageDataCollator` to:
    - propagate `answer_only_loss`, chat-template, and DFlash label-alignment settings;
    - apply `VLM_MIN_PIXELS` / `VLM_MAX_PIXELS` processor limits;
    - derive assistant-only masks from ChatML/Llama chat boundaries when processor generation masks are unavailable;                                                                                                                                
    - enforce the fixed `training_seq_len` required by DFlash block training.
  - Preserve the existing text-only DFlash path.                                                                                                                                                                                                    
                                                             
### Usage

```bash                                                                
python -m torch.distributed.run \                                      
--nproc_per_node 4 \                                                 
examples/speculative_decoding/main.py \                              
--config modelopt_recipes/general/speculative_decoding/dflash.yaml \
model.model_name_or_path=/path/to/qwen3-vl-model \
model.trust_remote_code=true \
data.data_path=/path/to/train.jsonl \
data.vlm_processor=/path/to/qwen3-vl-model \                        
data.vlm_img_dir=/path/to/image/root \
training.training_seq_len=4096 \
training.answer_only_loss=true \
dflash.dflash_block_size=8 \
dflash.dflash_mask_token_id=151669


### Testing
- git diff --check
- Parsed all modified Python modules successfully.
- Ran iterative multi-node Slurm smoke tests with a Qwen3-VL-family model and mixed multimodal data:
    - validated VLM model loading with Transformers 5;
    - validated distributed initialization, DFlash conversion, and VLM collation paths;
    - identified and addressed processor padding/truncation behavior required by fixed-size DFlash blocks.
This PR remains draft pending a completed end-to-end training smoke test and automated regression coverage.

### Before your PR is "Ready for review"
Make sure you read and follow Contributor guidelines (https://github.com/NVIDIA/Model-Optimizer/blob/main/CONTRIBUTING.md) and your commits are signed (git commit -s -S).
Make sure you read and follow the Security Best Practices (https://github.com/NVIDIA/Model-Optimizer/blob/main/SECURITY.md#security-coding-practices-for-contributors) (e.g. avoiding hardcoded trust_remote_code=True, torch.load(...,
weights_only=False), pickle, etc.).
- Is this change backward compatible?: ✅
- If you copied code from any other sources or added a new PIP dependency, did you follow guidance in CONTRIBUTING.md: N/A
- Did you write any new necessary tests?: ❌ — automated Qwen3-VL/DFlash regression coverage still needs to be added before review.
- Did you update Changelog (https://github.com/NVIDIA/Model-Optimizer/blob/main/CHANGELOG.rst)?: ❌ — evaluate and add an entry before marking ready for review if this is considered user-facing speculative-decoding support.
- Did you get Claude approval on this PR?: N/A
### Additional Information
The PR intentionally excludes local Slurm launch scripts, logs, model paths, datasets, and environment-specific configuration.


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Enhanced vision-language training with tokenizer/chat-template alignment, improved answer-only loss handling, and more consistent label shifting.
  * Expanded multimodal support with configurable VLM pixel-count bounds (via environment settings) to control processor expansion.

* **Bug Fixes**
  * More resilient JSONL dataset loading with an automatic-to-manual fallback for local, non-streaming inputs.
  * Improved training-time multimodal forward handling, including validation of hidden-state outputs.
  * More reliable distributed training behavior when no valid blocks are found, keeping gradient synchronization stable.

* **Tests**
  * Added coverage for Transformers VLM auto-model selection and multimodal collation label/mask correctness.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

### CODE REVIEW

File: modelopt/torch/speculative/plugins/hf_dflash.py [hunk 1]
```diff
@@ -575,12 +575,48 @@ def forward(
...
+            multimodal_keys = {
+                "pixel_values",
+                "pixel_values_videos",
+                "image_grid_thw",
+                "video_grid_thw",
+                "image_sizes",
+                "images",
+                "videos",
+            }
+            use_top_level_forward = any(
```
Comment (h-guo18) [inline modelopt/torch/speculative/plugins/hf_dflash.py:617]: LGTM for now. In the long term I think we want to abstract away the model-specific logics from the algorithm logic. Could you help add a TODO comment here?

**Review by h-guo18 — COMMENTED** [review]
Hi @skierat, thanks for the PR! The changes LGTM. Could you please:

-  Add some minimal unit tests for the changes in `VisionLanguageDataCollator` and `load_vlm_or_llm`
- Run a end2end test with `tools/launcher` and add a yaml example under `tools/launcher/examples`, similar to the existing `tools/launcher/examples/Qwen/Qwen3-8B/hf_online_dflash.yaml`. (Also works if you prefer to put in a separate PR).

This is a new place we put per-model examples. It should be much easier to create, test and share with users than our previous shell scripts or notebooks.

File: modelopt/torch/utils/plugins/transformers_dataset.py [hunk 2]
```diff
@@ -340,21 +398,204 @@ def __init__(
...
+
+        for key, pad_value in (
+            ("input_ids", self.tokenizer.pad_token_id),
+            ("attention_mask", 0),
+        ):
+            value = tokenized_messages.get(key)
+            if not isinstance(value, torch.Tensor):
+                continue
+            pad = value.new_full((value.shape[0], pad_width), pad_value)
+            tokenized_messages[key] = torch.cat((value, pad), dim=1)
```
[the remark this answers is not in the document]
Comment (h-guo18) [reply modelopt/torch/utils/plugins/transformers_dataset.py:545]: @skierat Please take a look if this make sense. Thanks

**Review by h-guo18 — APPROVED** [review]
lgtm. Approve to unblock; Please also take a look at AI review comments. Thanks!
